feat/tdd

Lainera opened the pull request:
Add tdd
Add Vagrantfile
Add playbook

Signed-off-by: Lainera <fistlain@hotmail.com>

Lainera committed:
feat/tdd
Add tdd
Add Vagrantfile
Add playbook

Signed-off-by: Lainera <fistlain@hotmail.com>

Lainera committed:
Update tls

Signed-off-by: Lainera <fistlain@hotmail.com>

r0mant commented on Readme.md:
```diff
@@ -0,0 +1,369 @@
+
+message Ack {}
+
+message JobId { bytes jobid = 1; }
```
What is used as a job ID?

Lainera replied:
Intention is to use [uuid](https://docs.rs/uuid/latest/uuid/), added comment.

r0mant commented on Readme.md:
```diff
@@ -0,0 +1,369 @@
+
+Adding process to a cgroup requires eid of the process performing the move to be the owner of common ancestor of source and destination cgroups. In other words, for child C of process P belonging to non-root user U, to be moved in a cgroup owned by U, P itself needs to be in a cgroup that is direct ancestor of target cgroup and owned by U.   
+This could be achieved by setting up appropriate permissions beforehand as privileged user and executing binary as U.
+For this implementation it is a no-goal and process is ran as root. 
```
Could you expand on your process execution model? It's not very clear from this description how the server will launch the jobs and place them into cgroups.

Lainera replied:
Added **Flow** section below, pls lmk if I'm missing something and/or any further clarifications are needed. 
Do mermaid diagrams open for you? If for some reason they do not - code could be pasted @[mermaid live editor](https://mermaid.live/)

r0mant commented on Readme.md:
```diff
@@ -0,0 +1,369 @@
+[![](https://mermaid.ink/img/eyJjb2RlIjoic2VxdWVuY2VEaWFncmFtXG5hY3RvciBDQSBhcyBDbGllbnQgQVxucGFydGljaXBhbnQgUyBhcyBKb2IgUnVubmVyIFNlcnZpY2UgXG5wYXJ0aWNpcGFudCBDUkEgYXMgQ29udHJvbGxlciBmb3IgQ2xpZW50IEFcbnBhcnRpY2lwYW50IEogYXMgSm9iXG5wYXJ0aWNpcGFudCBGUyBhcyBGaWxlIFN5c3RlbVxuXG5Ob3RlIG92ZXIgQ0EsRlM6IEpvYiB3aXRoIGlkIEpvYklkIGZvciBDbGllbnQgQSB3YXMgc3RhcnRlZCBlYXJsaWVyXG5KLSlGUzogU3RyZWFtcyBsb2dzXG5DQS ...
+
+Notes:
+- Considered keeping logs in-memory, but I believe this approach has couple of drawbacks. Reads and writes need to be synchronized somehow, so our options are either introducing Mutex, which may become fairly hot (depends on the volume of logs), or making Job own the output of the child process instead of streaming to file. This consequently means that all job output is now in main process memory ...
```
How are clients going to get the new output produced by the job in real-time?

Lainera replied:
Good callout, added a note on `inotify`

xacrimon commented:
LGTM once Roman's comments are sorted.

Lainera committed:
feat/tdd

Address CR questions:
	- add section on process flow
	- add uuid comment
	- add inotify section to clarify streaming part

Signed-off-by: Lainera <fistlain@hotmail.com>

r0mant commented on Readme.md:
```diff
@@ -0,0 +1,375 @@
+
+Notes:
+- Considered keeping logs in-memory, but I believe this approach has couple of drawbacks. Reads and writes need to be synchronized somehow, so our options are either introducing Mutex, which may become fairly hot (depends on the volume of logs), or making Job own the output of the child process instead of streaming to file. This consequently means that all job output is now in main process memory ...
+- Assuming job is still running, only logs that were generated at the moment of request would be sent back. To address that, controller could leverage [inotify](https://man7.org/linux/man-pages/man7/inotify.7.html) (intention is to use [inotify-rs](https://docs.rs/inotify/latest/inotify/)). Given log streamer task read EOF from open file, instead of exiting it `select!`'s on one of two events - e ...
```
> only logs that were generated at the moment of request would be sent back

Just wanted to clarify what this means exactly - will the client still receive all logs produced by the job from the beginning? That is part of the challenge spec requirements.

Lainera replied:
Yes this is correct, each output request gets logs from the beginning of the log file and continues receiving logs as they come in.
